File: ontology/src/main/scala/com/gravity/ontology/VirtuosoOntology.scala
package com.gravity.ontology

import scala.collection.JavaConversions._
import org.slf4j.LoggerFactory
import vocab.{URIType, NS}
import collection.mutable.ListBuffer
import org.openrdf.model.impl.URIImpl
import java.sql.{DriverManager, ResultSet}
import org.openrdf.rio.{RDFHandler, RDFFormat, Rio}
import com.gravity.ontology.VirtuosoOntology.TopicIterator
import org.openrdf.query.{QueryLanguage, BindingSet}
import org.openrdf.model._
import org.joda.time.DateTime
import java.io.{FileReader, BufferedReader}
import org.openrdf.repository.RepositoryException
import com.gravity.utilities.{Settings, Streams}
import scalaz.{Value =>_, _}
import Scalaz._

/**
 * User: chris
 * Date: 12/14/10
 * Time: 3:24 PM
 */

object VirtuosoOntology2 extends VirtuosoSparqlQuery {
  private implicit def bindingSetToRowWrapper(x: BindingSet) = new RowWrapper(x)

  implicit def nativeToResultSetWrapper(x: ResultSet) = new ResultSetWrapper(x)

  private val logger = LoggerFactory.getLogger(getClass)

  def getQuery(resName: String) = Streams.getResource("queries/" + resName, getClass)

  def alternateLabels(labelIterator: (URI, String) => Unit) = {
    query("Get alternate labels", "getAlternateLabels.query", (row, count) => {
      labelIterator(row.getURI("topicURI"), row.getString("alternateLabel"))
    })
  }

  def contextualPhrases(phraseIterator: (String, URI, Boolean) => Unit) = {
    query("Get contextual phrases", "getContextualPhrases.query", (row, count) => {
      val phraseRequired = row.getBool("phraseRequired")
      phraseIterator(row.getString("phraseText"), row.getURI("indicatedNode"), row.getBool("phraseRequired").get)
    })
  }

  def lowercaseBlacklist(blacklistIterator: (String) => Unit) = {
    query("Get lowercase blacklist", "getBlacklistLowercase.query", (row, count) => {
      blacklistIterator(row.getValue("blacklistedText").stringValue)
    })
  }

  def gravityOntology(rdfIterator: (URI, URI, Value) => Unit) = {
    query("Get gravity ontology", "getOntology.query", (row, count) => {
      rdfIterator(row.getURI("subject"), row.getURI("predicate"), row.getValue("object"))
    })
  }

  def topicTemplates(templateIterator: (URI, URI) => Unit) = {
    query("Get topic templates", "getTemplates.query", (row, count) => {
      templateIterator(row.getURI("topicId"), row.getURI("template"))
    })
  }

  def redirects(redirectIterator: (URI, URI) => Unit) = {
    query("Get redirects", "getRedirects.query", (row, count) => {
      redirectIterator(row.getURI("subject"), row.getURI("redirect"))
    })
  }

  def removals(removalIterator: (URI) => Unit) = {
    query("Get removals", "getRemovals.query", (row, count) => {
      removalIterator(row.getURI("topicId"))
    })
  }

  def renames(renameIterator: (URI, String) => Unit) = {
    query("Get renames", "getRenames.query", (row, count) => {
      renameIterator(row.getURI("topicId"), row.getString("renameTo"))
    })
  }

  def disambiguations(disIterator: (URI, URI) => Unit) = {
    query("Get disambiguations", "getDisambiguations.query", (row, count) => {
      disIterator(row.getURI("subject"), row.getURI("disambiguates"))
    })
  }

  def dbpediaOntology(ontologyIterator: (URI, String, Int, URI) => Unit) = {
    query("Get dbpedia ontology", "getDbPediaOntology.query", (row, count) => {
      if (!row.getString("name").contains("Aktivit")) {
        ontologyIterator(row.getURI("classId"), row.getString("name"), row.getInteger("level"), row.getURI("superClass"))
      }
    })
  }

  def topicConcepts(topicConceptIterator: (URI, URI) => Unit) = {
    query("Get topics to concepts", "getTopicConcepts.query", (row, count) => {
      if (NS.getType(row.getURI("topicId")) == URIType.TOPIC) {
        topicConceptIterator(row.getURI("topicId"), row.getURI("categoryId"))
      }
    })
  }

  def conceptHierarchy(conceptIterator: (URI, URI) => Unit) = {
    query("Get concept hierarchy", "getConceptHierarchy.query", (row, count) => {
      conceptIterator(row.getURI("conceptId"), row.getURI("targetId"))
    })
  }

  def concepts(conceptIterator: (URI, String) => Unit) = {
    query("Get concepts", "getAllConcepts.query", (row, count) => {
      conceptIterator(row.getURI("conceptId"), row.getString("conceptName"))
    })
  }

  def viewCounts(viewCountIterator: (URI, Int) => Unit) = {
    query("Get view counts", "getViewCounts.query", (row, count) => {
      viewCountIterator(row.getURI("subject"), row.getInteger("viewCount"))
    })
  }

  def topics(topicIterator: (URI, String) => Unit) = {
    nativequery("Get topics","getAllTopics.query",(row,count)=>{
      if (NS.getType(row.getURI("topicId")) == URIType.TOPIC) {
        topicIterator(row.getURI("topicId"), row.getString("topicName"))
      }
    })
  }

  def topicsFile(topicIterator:TopicIterator) = {
    val path = "/usr/local/virtuoso-opensource/var/lib/virtuoso/ontology_db/rdf/ontology/dbp36/labels_en.nt"
    rdfFile(path,st=>{
      topicIterator.row(st.getSubject.asInstanceOf[URI],st.getObject().stringValue,null,null)
    })
  }

  def rdfFile(path:String, handler:(Statement) => Unit) = {
    val parser = Rio.createParser(RDFFormat.NTRIPLES)
    parser.setRDFHandler(new RDFHandler() {
      def startRDF(){}
      def handleStatement(st: Statement) {handler(st)}
      def handleNamespace(prefix: String, uri: String) {}
      def handleComment(comment: String) {}
      def endRDF(){}
    })
    parser.parse(new BufferedReader(new FileReader(path)),"")
  }

  def topicsToClasses(ttcIterator: (URI, URI) => Unit) = {
    query("Topics to classes","getTopicTypeInfo.query", (row,count)=>{
      val topicId = row.getURI("topicId")
      val typeId = row.getURI("typeId")
      ttcIterator(topicId,typeId)
    })
  }

  def makeBindings(pairs: Option[(String, Value)]*): Map[String, Value] = {
    pairs.flatten.toMap
  }

  def query(queryName: String, queryFile: String, row: (BindingSet, Int) => Unit) {
    val connection = graph.getConnection
    try {
      val tupleQuery = connection.prepareTupleQuery(QueryLanguage.SPARQL, getQuery(queryFile))
      val queryResult = tupleQuery.evaluate
      var count = 0
      while (queryResult.hasNext) {
        count = count + 1
        row(queryResult.next, count)
        if (count % 10000 == 0) {
          logger.info(count + " rows through " + queryName)
        }
      }
      logger.info("Completed " + queryName + " with " + count + " rows.")
    } catch {
      case e: Exception => logger.error("Exception whilst issuing query", e)
    } finally {
      connection.close
    }
  }

  def withNativeConnection(stmtHandler: (java.sql.Statement) => Any)(implicit errorHandler: (Throwable) => Any = throw _) {
    Class.forName("virtuoso.jdbc3.Driver").newInstance
    val connection = DriverManager.getConnection(Settings.VIRTUOSO_LOCATION, Settings.VIRTUOSO_USERNAME, Settings.VIRTUOSO_PASSWORD)
    try {
      val stmt = connection.createStatement()
      stmtHandler(stmt)
    }
    catch {
      case ex: Exception => errorHandler(ex)
    }
    finally {
      connection.close()
    }
  }

  def nativequery(queryName: String, queryFile: String, row: (ResultSet, Int) => Unit) {
    var rowCount = 0
    //implicit val errorHandler = (ex: Throwable) => logger.error("Got exception at row " + rowCount, ex)
    withNativeConnection { stmt =>
      val more = stmt.execute("sparql " + getQuery(queryFile))
      if (more) {
        val resultSet = stmt.getResultSet
        while (resultSet.next) {
          rowCount = rowCount + 1
          row(resultSet, rowCount)
          if (rowCount % 10000 == 0) {
            logger.info(rowCount + " rows via " + queryName)
          }
        }
      }
    }
  }

  def main(args: Array[String]) = {

    this.topics((uri, name) => assert(uri != null && name != null))
  }

}

class RowWrapper(original: BindingSet) {
  def getURI(row: String): URI = {
    return original.getValue(row).asInstanceOf[URI]
  }

  def getBool(row: String) = Option(original.getValue(row)) map (_.asInstanceOf[Literal].booleanValue)

  def getString(row: String): String = {
    return original.getValue(row).stringValue
  }

  def getInteger(row: String): Int = {
    return original.getValue(row).asInstanceOf[Literal].intValue
  }

  def getDateTime(row: String): DateTime = {
    new DateTime(original.getValue(row).asInstanceOf[Literal].calendarValue().toGregorianCalendar)
  }

  def yomama() = {

  }

  override def toString = {
    original.getBindingNames map (name => name + "=" + original.getValue(name)) mkString (", ")
  }
}

class ResultSetWrapper(original: ResultSet) {
  def getURI(row: String): URI = {
    return new URIImpl(original.getString(row))
  }
}

trait SparqlQuery {
  def sparqlQuery(query: String, bindings: Map[String, Value] = Map.empty): Seq[RowWrapper]
}

trait IgnoreConnectionError extends SparqlQuery {
 import com.gravity.logging.Logging._
  abstract override def sparqlQuery(query: String, bindings: Map[String, Value] = Map.empty) = {
    try {
      super.sparqlQuery(query, bindings)
    }
    catch {
      case rex: RepositoryException if Option(rex.getCause) some (_.getMessage contains ("Connection refused")) none (false) => {
        warn("Ignoring connection error; returning empty result set. " + rex.getMessage)
        Seq.empty
      }
    }
  }
}

trait VirtuosoSparqlQuery extends SparqlQuery {
  protected val graph = RdfGraphFactory.getVirtuosoGraph

  private implicit def bindingSetToRowWrapper(x: BindingSet) = new RowWrapper(x)

  override def sparqlQuery(query: String, bindings: Map[String, Value] = Map.empty): Seq[RowWrapper] = {
    val connection = graph.getRepository.getConnection
    try {
      val tupleQuery = connection.prepareTupleQuery(QueryLanguage.SPARQL, query)
      bindings foreach { case (k, v) => tupleQuery.setBinding(k, v) }

      val queryResult = tupleQuery.evaluate
      val rows = new ListBuffer[RowWrapper]
      while (queryResult.hasNext) {
        rows += queryResult.next
      }

      rows
    }
    finally {
      connection.close()
    }
  }
}

File: ontology/src/main/java/com/gravity/ontology/VirtuosoOntology.java
package com.gravity.ontology; /**
 * User: chris
 * Date: Jul 18, 2010
 * Time: 1:02:54 PM
 */

import com.gravity.ontology.vocab.NS;
import com.gravity.ontology.vocab.URIType;
import com.gravity.utilities.Settings;
import com.gravity.utilities.Streams;
import org.openrdf.model.Literal;
import org.openrdf.model.URI;
import org.openrdf.model.Value;
import org.openrdf.model.impl.URIImpl;
import org.openrdf.query.*;
import org.openrdf.repository.RepositoryConnection;
import org.openrdf.repository.RepositoryException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class VirtuosoOntology {
  private static Logger logger = LoggerFactory.getLogger(VirtuosoOntology.class);

  public static final VirtuosoOntology Instance;
  private RdfGraph graph;

  public static final String query_getAllConcepts = Streams.getResource("queries/getAllConcepts.query", VirtuosoOntology.class);
  public static final String query_getConceptHierarchy = Streams.getResource("queries/getConceptHierarchy.query", VirtuosoOntology.class);
  public static final String query_getAllTopics = Streams.getResource("queries/getAllTopics.query", VirtuosoOntology.class);
  public static final String query_getTopicConcepts = Streams.getResource("queries/getTopicConcepts.query", VirtuosoOntology.class);
  public static final String query_getDbPediaOntology = Streams.getResource("queries/getDbPediaOntology.query", VirtuosoOntology.class);
  public static final String query_getTopicTypeInfo = Streams.getResource("queries/getTopicTypeInfo.query", VirtuosoOntology.class);
  public static final String query_getOntology = Streams.getResource("queries/getOntology.query", VirtuosoOntology.class);
  public static final String query_getViewCounts = Streams.getResource("queries/getViewCounts.query", VirtuosoOntology.class);
  public static final String query_getRedirects = Streams.getResource("queries/getRedirects.query", VirtuosoOntology.class);
  public static final String query_getDisambiguations = Streams.getResource("queries/getDisambiguations.query", VirtuosoOntology.class);
  public static final String query_getRemovals = Streams.getResource("queries/getRemovals.query", VirtuosoOntology.class);
  public static final String query_getRenames = Streams.getResource("queries/getRenames.query", VirtuosoOntology.class);
  public static final String query_getTemplates = Streams.getResource("queries/getTemplates.query", VirtuosoOntology.class);
  public static final String query_getContextualPhrases = Streams.getResource("queries/getContextualPhrases.query", VirtuosoOntology.class);
  public static final String query_getBlacklistLowercase = Streams.getResource("queries/getBlacklistLowercase.query", VirtuosoOntology.class);
  public static final String query_getAlternateLabels = Streams.getResource("queries/getAlternateLabels.query", VirtuosoOntology.class);

  public interface RdfIterator {
    public void row(URI subject, URI predicate, Value object);
  }

  public void iterateAlternateLabels(final AlternateLabelIterator iterator) {
    this.doQuery("Get alternate labels", query_getAlternateLabels, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        iterator.row((URI) row.getValue("topicURI"), row.getValue("alternateLabel").stringValue());
      }
    });
  }

  public void iterateContextualPhrases(final ContextualPhraseIterator iterator) {
    this.doQuery("Get contextual phrases", query_getContextualPhrases, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        boolean phraseRequired = ((Literal) row.getValue("phraseRequired")).booleanValue();
        iterator.row(row.getValue("phraseText").stringValue(), (URI) row.getValue("indicatedNode"), phraseRequired);
      }
    });
  }

  public void iterateLowercaseBlacklist(final SingleStringIterator iterator) {
    this.doQuery("Get blacklist lowercase", query_getBlacklistLowercase, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        String phrase = ((Literal) row.getValue("blacklistedText")).stringValue();
        iterator.row(phrase);
      }
    });
  }

  public void iterateGravityOntology(final RdfIterator gravityIterator) {
    this.doQuery("Get Gravity ontology", query_getOntology, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        gravityIterator.row((URI) row.getValue("subject"), (URI) row.getValue("predicate"), row.getValue("object"));
      }
    });
  }

  public void iterateTopicTemplates(final TemplateIterator templateIterator) {
    this.doQuery("Get templates", query_getTemplates, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        templateIterator.row((URI) row.getValue("topicId"), (URI) row.getValue("template"));
      }
    });
  }

  public void iterateRedirects(final RedirectIterator redirectIterator) {
    this.doQuery("Get redirects", query_getRedirects, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        redirectIterator.row((URI) row.getValue("subject"), (URI) row.getValue("redirect"));
      }
    });
  }

  public void iterateRemovals(final RemovalIterator removalIterator) {
    this.doQuery("Get removals", query_getRemovals, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        removalIterator.row((URI) row.getValue("topicId"));
      }
    });
  }

  public void iterateRenames(final RenameIterator renameIterator) {
    this.doQuery("Get renames", query_getRenames, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        renameIterator.row((URI) row.getValue("topicId"), row.getValue("renameTo").stringValue());
      }
    });
  }

  public void iterateDisambiguations(final DisambiguationIterator disambiguationIterator) {
    this.doQuery("Get disambiguations", query_getDisambiguations, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        disambiguationIterator.row((URI) row.getValue("subject"), (URI) row.getValue("disambiguates"));
      }
    });
  }


  public void iterateDbPediaOntology(final DbPediaIterator dbPediaIterator) {
    this.doQuery("Get dbpedia ontology", query_getDbPediaOntology, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        URI classId = (URI) row.getValue("classId");
        String name = row.getValue("name").stringValue();
        int level = ((Literal) row.getValue("level")).intValue();
        URI superClassId = null;
        if (row.getValue("superClass") != null) {
          superClassId = (URI) row.getValue("superClass");
        }
        if (name.contains("Aktivit")) {
        } else {
          dbPediaIterator.row(classId, name, level, superClassId);
        }
      }
    });
  }

  public void iterateTopicConcepts(final TopicToConceptIterator topicRelationshipIterator) {
    this.doQuery("Get topics to concepts", query_getTopicConcepts, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        URI topicURI = (URI) row.getValue("topicId");
        URI category = (URI) row.getValue("categoryId");
        if (NS.getType(topicURI) != URIType.TOPIC) {
          logger.info("Encountered URI that is not topic " + topicURI.toString());
        } else {
          topicRelationshipIterator.row(topicURI, category);
        }
      }
    });
  }

  /**
   * Provides a rowset that shows a concept and its broader concept
   *
   * @param conceptHierarchyIterator Implement this to get calls per row
   */
  public void iterateConceptHierarchy(final ConceptHierarchyIterator conceptHierarchyIterator) {
    doQuery("Get concept hierarchy", query_getConceptHierarchy, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        URI conceptURI = (URI) row.getValue("conceptId");
        URI broaderConceptURI = (URI) row.getValue("targetId");
        conceptHierarchyIterator.row(conceptURI, broaderConceptURI);
      }
    });

  }

  public void iterateConcepts(final ConceptIterator conceptIterator) {
    doNativeQuery("Get all concepts", query_getAllConcepts, new NativeRow() {
      @Override
      public void row(ResultSet row, int rowCount) throws SQLException {
        URI conceptURI = new URIImpl(row.getString("conceptId"));
        String conceptName = row.getString("conceptName");

        conceptIterator.row(conceptURI, conceptName);
      }

    });
  }

  public void iterateViewCounts(final ViewCountIterator viewCountIterator) {
    doQuery("Get view counts", query_getViewCounts, new Row() {
      @Override
      public void row(BindingSet row, int rowCount) {
        Value viewCount = row.getValue("viewCount");
        viewCountIterator.row((URI) row.getValue("subject"), ((Literal) viewCount).intValue());
      }
    });
  }

  public void iterateTopics(final TopicIterator topicIterator) {

    doNativeQuery("Get all topics",query_getAllTopics,new NativeRow() {
      @Override
      public void row(ResultSet rs, int rowCount) throws SQLException {
        String topicId = rs.getString("topicId");
        String topicName = rs.getString("topicName");
        if(topicId != null && topicName != null) {
          URI topicURI = new URIImpl(topicId);
          if(NS.getType(topicURI) != URIType.TOPIC) {
            logger.info("Encountered URI that is not topic " + topicURI.toString());
          }else {
            topicIterator.row(topicURI,topicName,null,null);
          }

        }
      }
    });
  }

  public void iterateTopicsToClasses(final TopicToClassIterator topicToClassIterator) {
    doQuery("Topics to Classes", query_getTopicTypeInfo, new Row() {
      URI currentTopic = null;
      List<URI> classes = null;
      int relationshipsCreated = 0;

      @Override
      public void row(BindingSet row, int rowCount) {
        boolean logit = logger.isDebugEnabled();
        URI topicId = (URI) row.getValue("topicId");
        URI typeId = (URI) row.getValue("typeId");
        String topicName = row.getValue("topicName").stringValue();
        boolean topicIsResource = topicId.getNamespace().equals(NS.DBPEDIA_TOPIC_NAMESPACE);
        if (!topicIsResource) {
          if (logit) logger.debug("topic " + topicId.toString() + " is not dbpedia resource, skipping");
        } else {
          if (logit) logger.debug("Matching " + topicId + " against type " + typeId);
          if (currentTopic == null || !currentTopic.equals(topicId)) {
            if (logit) logger.debug("Initializing " + topicId);
            if (currentTopic != null) {
              //THIS IS THE POINT WHERE WE SAVE THE CURRENTTOPIC AND TYPEBUFFER TO THE INDEX
              topicToClassIterator.row(currentTopic, topicName, classes);
              //END POINT WHERE WE SAVE TO INDEX
            }
            currentTopic = topicId; //initialize
            classes = new ArrayList<URI>();
            classes.add(typeId);
          } else {
            classes.add(typeId);
          }
        }
      }


    });
  }

  static {
    Instance = new VirtuosoOntology(RdfGraphFactory.getVirtuosoGraph());
  }

  public interface DbPediaIterator {
    public void row(URI classId, String name, int level, URI superClassId);
  }

  /**
   * Interface for clients who wish to provide a safe row evaluator
   */
  public interface Row {
    void row(BindingSet row, int rowCount);
  }

  public void doNativeQuery(String queryName, String queryText, NativeRow row) {
    Connection connection = null;
    try {
      Class.forName("virtuoso.jdbc3.Driver").newInstance();
      connection = DriverManager.getConnection(Settings.VIRTUOSO_LOCATION, Settings.VIRTUOSO_USERNAME, Settings.VIRTUOSO_PASSWORD);

      Statement stmt = connection.createStatement();
      boolean more = stmt.execute("sparql " + queryText);

      if (more) {
        ResultSet rs = stmt.getResultSet();
        int rowCount = 0;
        while (rs.next()) {
          rowCount++;
          row.row(rs, rowCount);
          if (rowCount % 10000 == 0) {
            logger.info(rowCount + " native virtuoso rows through " + queryName);
          }
        }
      }
    } catch (Exception exc) {
      throw new RuntimeException(exc);
    } finally {
      if (connection != null) {
        try {
          connection.close();
        } catch (SQLException e) {
          logger.error("Exception", e);
        }
      }
    }
  }

  public void doQuery(String queryName, String queryText, Row row) {
    RepositoryConnection connection = graph.getConnection();
    try {
      TupleQuery query = connection.prepareTupleQuery(QueryLanguage.SPARQL, queryText);
      TupleQueryResult queryResult = query.evaluate();
      int count = 0;
      while (queryResult.hasNext()) {
        count++;
        BindingSet bset = queryResult.next();
        row.row(bset, count);
        if (count % 10000 == 0) {
          logger.info(count + " rows through " + queryName);
        }
      }
      connection.close();
    } catch (MalformedQueryException e) {
      throw new RuntimeException(e);
    } catch (RepositoryException e) {
      throw new RuntimeException(e);
    } catch (QueryEvaluationException e) {
      throw new RuntimeException(e);
    }
  }


  private VirtuosoOntology(RdfGraph graph) {
    this.graph = graph;
  }


  public interface NativeRow {
    public void row(ResultSet rs, int rowCount) throws SQLException;
  }

  public interface SingleStringIterator {
    public void row(String str);
  }


  public interface TopicToConceptIterator {
    public void row(URI topicURI, URI category);
  }

  public interface ConceptHierarchyIterator {
    public void row(URI conceptURI, URI broaderConceptURI);
  }

  public interface RedirectIterator {
    public void row(URI topicURI, URI redirectURI);
  }

  public interface DisambiguationIterator {
    public void row(URI topicURI, URI disambiguates);
  }


  public interface ConceptIterator {
    public void row(URI conceptURI, String conceptName);
  }

  public interface TopicIterator {
    public void row(URI topicURI, String topicName, String shortAbstract, String longAbstract);
  }

  public interface TopicToClassIterator {
    public void row(URI currentTopic, String topicName, List<URI> classes);
  }

  public interface ViewCountIterator {
    public void row(URI currentTopic, int viewCount);

  }

  public interface RemovalIterator {
    public void row(URI uriToRemove);
  }

  public interface RenameIterator {
    public void row(URI uriToRename, String renameTo);
  }

  public interface TemplateIterator {
    public void row(URI topic, URI template);
  }

  public interface ContextualPhraseIterator {
    public void row(String phraseText, URI targetNode, boolean isRequired);
  }

  public interface AlternateLabelIterator {
    public void row(URI topicURI, String label);
  }
}
